### PR TITLE
Feature/zr300 for torso cameras

### PR DESCRIPTION
**Pull request by floweisshardt**
~~requires https://github.com/ipa320/cob_calibration_data/pull/136~~

requires https://github.com/ipa320/cob_common/pull/225
requires https://github.com/intel-ros/realsense/pull/244 to be merged

though minor, issue related https://github.com/intel-ros/realsense/issues/258

### CODE REVIEW

File: cob_bringup/components/cam3d_zr300_rgbd.launch [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+<launch>
+	<arg name="robot"/>
+	<arg name="name"/>
+	<arg name="num_worker_threads" default="4"/>
+	<arg name="start_manager" default="true"/>
+	<arg name="flip" default="true"/>
+	<arg name="sim" default="false"/>
+	<arg name="serial_no" default=""/>
+
+	<include unless="$(arg sim)" file="$(find realsense_camera)/launch/zr300_nodelet_rgbd.launch">
```
Comment (fmessmer) [inline cob_bringup/components/cam3d_zr300_rgbd.launch:11]: do you have an overlay of `realsense_camera`?
in my `/opt/ros/` I can only find a `zr300_nodelet_default.launch` file

File: cob_bringup/components/cam3d_zr300_rgbd.launch [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	<include if="$(arg flip)" file="$(find cob_bringup)/drivers/image_flip_nodelet.launch">
+		<arg name="robot" value="$(arg robot)"/>
+		<arg name="camera_name" value="$(arg name)"/>
+		<arg name="nodelet_manager" value="$(arg name)_nodelet_manager"/>
+		<arg name="start_manager" value="$(arg start_manager)"/>
+	</include>
+
+
+	<group ns="$(arg name)" if="$(arg sim)">
+		<!-- static frames broadcasted by camera driver -->
```
Comment (fmessmer) [inline cob_bringup/components/cam3d_zr300_rgbd.launch:30]: the values below are for copy-pasted from `cob_bringup/components/cam3d_r200_rgbd.launch`, right?
we need to verify/get/update these transforms with a zr300!

File: cob_bringup/components/cam3d_zr300_rgbd.launch [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		<node pkg="tf2_ros" type="static_transform_publisher" name="stp_depth_frame" args="0.0 0.059 0.0 0.0 0.0 0.0 1.0 /$(arg name)_link /$(arg name)_depth_frame"/>
+		<node pkg="tf2_ros" type="static_transform_publisher" name="stp_depth_optical_frame" args="0.0 0.0 0.0 -0.5 0.5 -0.5 0.5 /$(arg name)_depth_frame /$(arg name)_depth_optical_frame"/>
+
+		<node pkg="tf2_ros" type="static_transform_publisher" name="stp_ir_frame" args="0.0 0.059 0.0 0.0 0.0 0.0 1.0 /$(arg name)_link /$(arg name)_ir_frame"/>
+		<node pkg="tf2_ros" type="static_transform_publisher" name="stp_ir_optical_frame" args="0.0 0.0 0.0 -0.5 0.5 -0.5 0.5 /$(arg name)_ir_frame /$(arg name)_ir_optical_frame"/>
+
+		<node pkg="tf2_ros" type="static_transform_publisher" name="stp_ir2_frame" args="0.0 -0.011 0.0 0.0 0.0 0.0 1.0 /$(arg name)_link /$(arg name)_ir2_frame"/>
+		<node pkg="tf2_ros" type="static_transform_publisher" name="stp_ir2_optical_frame" args="0.0 0.0 0.0 -0.5 0.5 -0.5 0.5 /$(arg name)_ir2_frame /$(arg name)_ir2_optical_frame"/>
+
+		<!-- relay additional topics not available from sensor plugin -->
```
Comment (fmessmer) [inline cob_bringup/components/cam3d_zr300_rgbd.launch]: also, the topics need to be compared between sim and hw

File: cob_bringup/robots/cob4-7.xml [hunk 4]
```diff
@@ -128,11 +128,9 @@
 			<arg name="sim" value="$(arg sim)"/>
 			<arg if="$(arg sim)" name="cartesian_control" value="true"/>
 		</include>
-		<include file="$(find cob_bringup)/components/cam3d_openni2.launch">
+		<include file="$(find cob_bringup)/components/cam3d_zr300_rgbd.launch">
 			<arg name="robot" value="$(arg robot)"/>
-			<arg name="name" value="torso_cam3d_right"/>
-			<arg name="device_id" value="#1"/>
-			<arg name="data_skip" value="0"/>
+			<arg name="name" value="torso_cam3d_left"/>
```
Comment (fmessmer) [inline cob_bringup/robots/cob4-7.xml:133]: why has the sides been swapped?
cam_left is now on t1 - cam_right on t3
the other robots have it the other way! can we re-plug the cameras?

File: cob_bringup/robots/cob4-7.xml [hunk 5]
```diff
@@ -201,12 +200,9 @@
...
-		<include file="$(find cob_bringup)/components/cam3d_openni2.launch">
+		<include file="$(find cob_bringup)/components/cam3d_zr300_rgbd.launch">
 			<arg name="robot" value="$(arg robot)"/>
-			<arg name="name" value="torso_cam3d_left"/>
-			<arg name="device_id" value="#1"/>
-			<arg name="data_skip" value="0"/>
+			<arg name="name" value="torso_cam3d_right"/>
 			<arg name="num_worker_threads" value="2"/>
 			<arg name="start_manager" value="false"/>
 			<arg name="sim" value="$(arg sim)"/>
```
Comment (fmessmer) [inline cob_bringup/robots/cob4-7.xml:207]: this camera has not `serial_no` argument!

File: cob_hardware_config/robots/cob4-7/urdf/cob4-7.urdf.xacro [hunk 6]
```diff
@@ -24,6 +24,7 @@
 
   <!-- sensors -->
   <xacro:include filename="$(find cob_description)/urdf/sensors/realsense.urdf.xacro" />
+  <xacro:include filename="$(find cob_description)/urdf/sensors/realsense_zr300.urdf.xacro" />
```
Comment (fmessmer) [inline cob_hardware_config/robots/cob4-7/urdf/cob4-7.urdf.xacro:27]: this file does not exist yet! the PR to cob_common is missing!

BTW, what's the difference between the two urdf descriptions? if there is none (or it is minimal), I'd prefer to have both realsense_cameras in the same xacro....(and maybe use some xacro arguments)

**Review by fmessmer — CHANGES_REQUESTED** [review]

File: cob_bringup/components/cam3d_zr300_rgbd.launch (hunk 1 above)
Comment (floweisshardt) [reply cob_bringup/components/cam3d_zr300_rgbd.launch:11]: yes we need an overlay. ZR300 launch files are already merged but not released again.

File: cob_bringup/robots/cob4-7.xml (hunk 4 above)
Comment (floweisshardt) [reply cob_bringup/robots/cob4-7.xml:133]: this way it's more intuitive: left pc + left camera and right pc + right camera

File: cob_hardware_config/robots/cob4-7/urdf/cob4-7.urdf.xacro (hunk 6 above)
Comment (floweisshardt) [reply cob_hardware_config/robots/cob4-7/urdf/cob4-7.urdf.xacro:27]: PR in cob_common is https://github.com/ipa320/cob_common/pull/225

at the moment the files are copied. But as soon as we harmonize hw and sim, we need to add the fisheye camera and if possible the IMU which is not available on the r200

File: cob_bringup/components/cam3d_zr300_rgbd.launch (hunk 1 above)
Comment (fmessmer) [reply cob_bringup/components/cam3d_zr300_rgbd.launch:11]: then please add the repo to the rosinstall file for Travis CI to pass!

Comment (fmessmer) [reply cob_bringup/components/cam3d_zr300_rgbd.launch:11]: this is not even an unreleased version of official `intel-ros/realsense` but a [feature branch](https://github.com/ipa-bnm/realsense/tree/fix/launch_file_rename) of `ipa-bnm/realsense` see https://github.com/intel-ros/realsense/pull/244

File: cob_bringup/components/cam3d_zr300_rgbd.launch (hunk 2 above)
Comment (fmessmer) [reply cob_bringup/components/cam3d_zr300_rgbd.launch:30]: values are correct for zr300 as well
still frame_ids are wrong (not prefixed correctly by driver see https://github.com/intel-ros/realsense/issues/258

Comment (fmessmer) [reply cob_bringup/components/cam3d_zr300_rgbd.launch:30]: missing static transforms are added in ipa-fmw/cob_robots#45

File: cob_bringup/robots/cob4-7.xml (hunk 5 above)
Comment (benmaidel) [reply cob_bringup/robots/cob4-7.xml:207]: right there should not be a serial_no arguemnt

**Review by benmaidel — DISMISSED** [review]

File: cob_bringup/components/cam3d_zr300_rgbd.launch (hunk 1 above)
Comment (benmaidel) [reply cob_bringup/components/cam3d_zr300_rgbd.launch:11]: added realsense to travis rosinstall

File: cob_bringup/components/cam3d_zr300_rgbd.launch (hunk 2 above)
Comment (benmaidel) [reply cob_bringup/components/cam3d_zr300_rgbd.launch:30]: already merged!

**Review by benmaidel — DISMISSED** [review]

**Comment by floweisshardt** [comment]
@ipa-fxm: please review again

File: cob_bringup/components/cam3d_zr300_rgbd.launch (hunk 1 above)
Comment (benmaidel) [reply cob_bringup/components/cam3d_zr300_rgbd.launch:11]: If we like to release cob_robots we should fall back to the typo rgdb, so we can use the released realsense package.

**Comment by fmessmer** [comment]
@ipa-jba @ipa-rmb @ipa-fez FYI
I don't want to approve this without your opinion

this is "being admin" at its finest :wink:

File: cob_bringup/components/cam3d_zr300_rgbd.launch (hunk 1 above)
Comment (fmessmer) [reply cob_bringup/components/cam3d_zr300_rgbd.launch:11]: even with switching back to rg**db** launch file...this is still only available in realsense_camera from source as it is not yet released...in the current release of `realsense_camera` there is only a `zr300_nodelet_default.launch` file for the zr300

**Review by jabbenseth — CHANGES_REQUESTED** [review]
until intel-ros/realsense#244 is merged and released

**Comment by fmessmer** [comment]
a fix for the issue with the optical frame prefixing (intel-ros/realsense#258) is proposed in intel-ros/realsense#259

**Comment by fmessmer** [comment]
rebase in #715

**Comment by fmessmer** [comment]
closing in favor of #715

**Pull request closed without merging**